File: packages/canvas/src-native/canvas-ios/TextBaseLine.swift
//
//  TextBaseLine.swift
//  CanvasNative
//
//  Created by Osei Fortune on 11/11/2020.
//

import Foundation

@objc(TNSTextBaseLine)
public enum TNSTextBaseLine: Int, RawRepresentable {
    case Top
    case Hanging
    case Middle
    case Alphabetic
    case Ideographic
    case Bottom
    
    public typealias RawValue = UInt32
    
    public var rawValue: RawValue {
        switch self {
        case .Top:
            return 0
        case .Hanging:
            return 1
        case .Middle:
            return 2
        case .Alphabetic:
            return 3
        case .Ideographic:
            return 4
        case .Bottom:
            return 5
        }
    }
    
    
    public init?(rawValue: RawValue) {
        switch rawValue {
        case 0:
            self = .Top
        case 1:
            self = .Hanging
        case 2:
            self = .Middle
        case 3:
            self = .Alphabetic
        case 4:
            self = .Ideographic
        case 5:
            self = .Bottom
        default:
            return nil
        }
    }
    
    
    public init?(string: String) {
        switch string {
        case "top":
            self = .Top
        case "hanging":
            self = .Hanging
        case "middle":
            self = .Middle
        case "alphabetic":
            self = .Alphabetic
        case "ideographic":
            self = .Ideographic
        case "bottom":
            self = .Bottom
        default:
            return nil
        }
    }
    
}

File: packages/canvas/src-native/canvas-native/canvas-core/src/common/context/text_styles/text_baseline.rs
#[repr(C)]
#[derive(Copy, Clone, Debug)]
pub enum TextBaseLine {
    TOP = 0,
    HANGING = 1,
    MIDDLE = 2,
    ALPHABETIC = 3,
    IDEOGRAPHIC = 4,
    BOTTOM = 5,
}

impl Default for TextBaseLine {
    fn default() -> Self {
        TextBaseLine::ALPHABETIC
    }
}

impl From<i32> for TextBaseLine {
    fn from(value: i32) -> TextBaseLine {
        match value {
            0 => Self::TOP,
            1 => Self::HANGING,
            2 => Self::MIDDLE,
            3 => Self::ALPHABETIC,
            4 => Self::IDEOGRAPHIC,
            5 => Self::BOTTOM,
            _ => TextBaseLine::ALPHABETIC
        }
    }
}

impl Into<i32> for TextBaseLine {
    fn into(self) -> i32 {
        match self {
            TextBaseLine::TOP => 0,
            TextBaseLine::HANGING => 1,
            TextBaseLine::MIDDLE => 2,
            TextBaseLine::ALPHABETIC => 3,
            TextBaseLine::IDEOGRAPHIC => 4,
            TextBaseLine::BOTTOM => 5
        }
    }
}